### DISCUSSION TITLE
Are there any new v2 features we might expect prior to the v1 shutdown?

### DISCUSSION THREAD

**Discussion by unitof** [vercel/vercel#4004] [General] [answered]
An email was sent out today announcing the shutdown of all Now v1 instances. We've been warned of this for a while and it makes sense—I'm not complaining.

From the aforementioned email:

>### Important Dates
>- 3rd of April 2020: ZEIT announces the deprecation of the ZEIT Now 1.0 platform.
>- 1st of May 2020: Creating new ZEIT Now 1.0 deployments will be prevented.
>- 7th of August 2020: All ZEIT Now 1.0 deployments will be deleted.

But v2 _was and is_ significantly different in design and purpose than v2 there are still things possible in v1 that are not compatible with v2.

The email suggests the following for migrating **continuously running servers**:

> ZEIT Now 2.0 does not support running servers. We recommend deploying these types of projects to a platform with support for servers.

Again: totally makes sense.

I'm starting this thread to ask if there might be any features on the v2 roadmap prior to August which _might_ make it possible to migrate v1 instances that are currently un-migratable, which might make it worth waiting a bit before migrating to a third party platform.

### The features that were possible in v1 but not yet in v2 that come to mind for me are:
- [ ] [Native Websocket support](https://github.com/zeit/docs/pull/1612)
- [ ] [Native cron/scheduled task support](https://github.com/zeit/now/issues/146)
  - Beloved hometown Homestar @styfle [said](https://github.com/zeit/now/issues/146#issuecomment-472450023) this was in the works as of last year
- [ ] [Larger lambda sizes](https://spectrum.chat/zeit/now/could-someone-point-me-toward-strategies-for-decreasing-my-php-lambda-size~635fb5c1-98dc-44ed-8574-41986487524b)
  - I'm gonna _guess_ this one's out of the question as large lambdas are antithetical to the v2 architecture—just putting it here since it's one thing currently preventing me from smoothly migrating one of my v1 instances.

Maybe I missed something though and one or more of these are now supported. If so, please let me know!

If there are other singular features preventing folks from migrating a v1 instance v2, post 'em here.

Thanks for the through and beautiful work on v2!

v1 brought me here, and I'll remember it fondly.

**Comment by simonw** [comment]
I really need support for the Python standard library sqlite3 module so I can query read-only SQLite binary files that I bundle with my deployments. Last time I checked that module wasn't available, I think because AWS lambda mistakenly thought that no one would want it in a serverless writable-filesystem-lacking environment.

**Reply by styfle** [reply]
Hi @simonw 

It is probably possible to copy the sqlite binary (I think you only need `_sqlite3.so`) into the Serverless Function. This is probably a brittle solution but it could be a workaround for your issue. Take a look at [sqlite-lambda-layer](https://github.com/dschep/sqlite-lambda-layer).

**Reply by simonw** [reply]
https://now-2-python-versions-ks69olzpi.now.sh/api - I just did an experiment and it looks like `sqlite3` may be there after all now!

It's 3.7.17 which is a pretty old version (2013 I think) but it might just work for me.

**Reply by styfle** [reply]
Recently @dnk8n [mentioned](https://github.com/zeit/now/issues/2979#issuecomment-609366369) that Python 3.8 ships with sqlite.

If we added support for Python 3.8 would that solve your problem?

**Reply by dnk8n** [reply]
Zeit's Python runtime does support sqlite. Here is a project in action https://covid-19-backend.now.sh.

Python 3.7 or 3.8 would be very welcome because of many features in the newer sqlite that fit very much in line with the serverless way of doing things. For example, easily loading a sqlite file into memory for faster queries (good idea for read-only applications).

Allowing anything but non-persistent writes (I would actually go so far as to say, it needs to be read-only) to a sqlite database deployed on Zeit would be unusable anyway.

**Reply by dnk8n** [reply]
[Off-topic, feel free to ignore] A weird thing I noticed when things are deployed to Zeit, the creation and modification times of files are way off. Not entirely sure why that would be! This can be seen in the link above.

**Reply by simonw** [reply]
I managed to get a mostly-working demo of Datasette shipped here! https://now-2-datasette-e4l4whc9f.now.sh/

There are a few problems with it - see ongoing issue thread here: https://github.com/simonw/datasette/issues/717 - and it would definitely benefit from a more recent SQLite3 version that supports FTS5. Promising start though.

**Reply by unitof** [reply]
@simonw Neat! Would you be willing to [set `public: true`](https://zeit.co/docs/configuration#project/public) in `now.json` to open up `/_src` & `_logs`? Curious to see.

**Reply by simonw** [reply]
@unitof I added a `--public` flag to the `datasette publish now ...` command that does that - here's an example: https://datasette-public.now.sh/_src

**Reply by simonw** [reply]
Also useful to know: it turns out you CAN upgrade the SQLite version available to Python on Now v3 using the [pysqlite3-binary](https://pypi.org/project/pysqlite3-binary/) PyPI module.

My `datasette-publish-now` plugin now uses that package: https://github.com/simonw/datasette-publish-now/issues/17

**Comment by paulogdm** [comment, marked as answer]
Negative. We don't plan to release those features even if some of them are being worked on.

> Websocket support

We recommend using pub'n'sub services like Pusher.

>  Native cron/scheduled task support

This is indeed in the works and planned. We are focusing on more features first, such as "multi-environment" workflows.

>  Larger lambda sizes

Unfortunately, not possible. This is an upstream provider hard limitation.
